### DISCUSSION TITLE
Build error under Archlinux

### DISCUSSION THREAD

**Discussion by UVJkiNTQ** [Difegue/LANraragi#1149] [Q&A] [answered]
**LRR Version and OS**
LANraragi 0.9.30
Archlinux 6.11.5-arch1-1

**Bug Details**
Build error in installing Perl Modules.

How to reproduce this error:
1. follow the instructions on https://sugoi.gitbook.io/lanraragi/dev/installing-lanraragi/community#arch-linux
2. replace the tarball version and relative hashsum to 0.9.30 in PKGBUILD and run `makepkg -rsi`

**Matching Logs**

> Installing Perl modules... This might take a while.
> 
> Installing Linux::Inotify2 for non-macOS systems... (This will do nothing if the package is there already)
> Linux::Inotify2 not installed! Trying to install now using cpanm
> Linux::Inotify2 is up to date. (2.3)
> --> Working on ./tools/.
> Configuring /home/[xxx]/pkgbuild/lanraragi/src/LANraragi-v.0.9.30/tools ... OK
> ==> Found dependencies: Authen::Passphrase
> --> Working on Authen::Passphrase
> Fetching https://www.cpan.org/authors/id/Z/ZE/ZEFRAM/Authen-Passphrase-0.008.tar.gz ... OK
> Configuring Authen-Passphrase-0.008 ... OK
> ==> Found dependencies: Crypt::DES
> --> Working on Crypt::DES
> Fetching https://www.cpan.org/authors/id/D/DP/DPARIS/Crypt-DES-2.07.tar.gz ... OK
> Configuring Crypt-DES-2.07 ... OK
> Building Crypt-DES-2.07 ... FAIL
> ! Installing Crypt::DES failed. See /home/[xxx]/.cpanm/work/1735662862.1758071/build.log for details. Retry with --force to force install it.
> ! Installing the dependencies failed: Module 'Crypt::DES' is not installed
> ! Bailing out the installation for Authen-Passphrase-0.008.
> ! Installing the dependencies failed: Module 'Authen::Passphrase' is not installed
> ! Bailing out the installation for ./tools/..
> Something went wrong while installing Perl modules - Bailing out. at ./tools/install.pl line 150.
> ==> ERROR: A failure occurred in build().
>     Aborting...

**Screenshots**

**Comment by Difegue** [comment]
I don't use Arch so I'm afraid I don't have many pointers here - It's probably worth checking the cpan build log, you might just be missing a crypto library here.

**Comment by UVJkiNTQ** [comment]
I am sorry for that. Here is the build log.
Since I am not familiar with Perl, I just put the log here to see if anyone else could help.

PKGBUILD:
<details>
#Maintainer: SadaoMaou <luxrayguy@gmail.com>
#Contributor: Satiricon <david.satiricon@gmail.com>
#Contributor: Alexbrenvero <alexandre@lempire.xyz>

pkgname=lanraragi
pkgver=v.0.9.30
pkgrel=1
pkgdesc="LANraragi build package"
arch=(any)
url="https://github.com/Difegue/LANraragi"
license=("MIT")
depends=(libarchive ghostscript perl gnupg redis imagemagick libwebp openssl zlib perl-local-lib perl-uri)
makedepends=(cpanminus npm perl-config-autoconf pkgconf)
source=("https://github.com/Difegue/LANraragi/archive/${pkgver}.tar.gz"
        "lanraragi.service" "lanraragi.sysusers" "lanraragi.tmpfiles")
sha512sums=(
"6f552d8eb50b6d0b5b66d91d6ee7dbc2ffb835ce1ead6e517a61c3e961e1b8156be9d4e871a8910c8d8215969d44e27ab225da2e00ec7ddda0232d7fdba7e3b2"
"1ef2ca60e51269351440c1ae77431c46b8b82eb7d285e6d209ca9ac64141e2a337a5e9abca49425d026a10194b341c6c22966708bfa8f81d3904c8bc490123e6"
"c598b37c691b66c3c32aed50d0e79c9d75708c40f79fe83f287e6ed1592736f608b646427ef044bf489d034a91219fd1567b3d0ece320633176ae5dbe28b7685"
"0f66197d8fe253d1f6ff56f4e301c337be66aa82cf091619df1a8c17ec97d66721f2618cdcdf88718f28f0d80263d4331e482d22468e1bd7bd55ecb2bc076a2d"
)


build() {
  PATH=/usr/bin/vendor_perl:$PATH
  
  cd LANraragi-${pkgver}
  export PERL_CPANM_OPT="--local-lib=./perl5 $PERL_CPANM_OPT"
  npm run lanraragi-installer install-full
}


package() {
  install -d -m 755 "${pkgdir}/usr/lib/lanraragi/perl5"
  cp -dpr --no-preserve=ownership "${srcdir}/LANraragi-${pkgver}/"* "${pkgdir}/usr/lib/lanraragi"
  chmod -R a=,a+rX,u+w ${pkgdir}/usr/lib/lanraragi

  install -D -m 644 "${srcdir}/lanraragi.service" "${pkgdir}/usr/lib/systemd/system/lanraragi.service"
  install -D -m 644 "${srcdir}/lanraragi.sysusers" "${pkgdir}/usr/lib/sysusers.d/lanraragi.conf"
  install -D -m 644 "${srcdir}/lanraragi.tmpfiles" "${pkgdir}/usr/lib/tmpfiles.d/lanraragi.conf"
  install -D -m 644 "${srcdir}/LANraragi-${pkgver}/COPYING" "$pkgdir/usr/share/licenses/$pkgname/LICENSE"

  grep -rnwl ${pkgdir} -e ${srcdir} | xargs -I % sed -i 's,'"$srcdir"',\.,g' %

}

</details>


Build log:
<details>

cpanm (App::cpanminus) 1.7047 on perl 5.040000 built for x86_64-linux-thread-multi
Work directory is /home/[xxx]/.cpanm/work/1735662862.1758071
You have make /usr/bin/make
You have /usr/bin/wget
You have /usr/bin/tar: tar (GNU tar) 1.35
Copyright (C) 2023 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <https://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by John Gilmore and Jay Fenlason.
You have /usr/bin/unzip
--> Working on ./tools/.
Entering /home/[xxx]/pkgbuild/lanraragi/src/LANraragi-v.0.9.30/tools
Configuring /home/[xxx]/pkgbuild/lanraragi/src/LANraragi-v.0.9.30/tools
-> OK
Checking if you have Compress::Zlib 2.087 ... Yes (2.212)
Checking if you have Archive::Libarchive::Extract 0.03 ... Yes (0.03)
Checking if you have URI::Escape 1.74 ... Yes (5.31)
Checking if you have Test::Deep 1.13 ... Yes (1.204)
Checking if you have Module::Pluggable 5.2 ... Yes (6.2)
Checking if you have IO::Socket::Socks 0.74 ... Yes (0.74)
Checking if you have Mojolicious::Plugin::Status 1.15 ... Yes (1.17)
Checking if you have Encode 3.01 ... Yes (3.21)
Checking if you have Sys::CpuAffinity 1.12 ... Yes (1.12)
Checking if you have Digest::SHA 6.02 ... Yes (6.04)
Checking if you have Parallel::Loops 0.1 ... Yes (0.12)
Checking if you have File::ReadBackwards 1.05 ... Yes (1.06)
Checking if you have YAML::PP v0.38.0 ... Yes (v0.38.0)
Checking if you have Test::MockObject 1.20200122 ... Yes (1.20200122)
Checking if you have Minion 10.31 ... Yes (10.31)
Checking if you have local::lib 2.000024 ... Yes (2.000029)
Checking if you have Mojolicious 9.39 ... Yes (9.39)
Checking if you have Mojolicious::Plugin::RenderFile 0.12 ... Yes (0.12)
Checking if you have Archive::Libarchive::Peek 0.04 ... Yes (0.05)
Checking if you have Logfile::Rotate 1.04 ... Yes (1.05)
Checking if you have Time::Local 1.3 ... Yes (1.35)
Checking if you have Sort::Naturally 1.03 ... Yes (1.03)
Checking if you have Locale::Maketext 1.33 ... Yes (1.33)
Checking if you have Proc::Simple 1.32 ... Yes (1.32)
Checking if you have File::ChangeNotify 0.31 ... Yes (0.31)
Checking if you have Locale::Maketext::Lexicon 1 ... Yes (1.00)
Checking if you have Crypt::Rijndael 1.14 ... Yes (1.16)
Checking if you have List::MoreUtils 0.43 ... Yes (0.430)
Checking if you have IPC::Cmd 1.02 ... Yes (1.04)
Checking if you have Test::Trap v0.3.4 ... Yes (0.3.5)
Checking if you have Test::Harness 3.42 ... Yes (3.48)
Checking if you have Mojolicious::Plugin::TemplateToolkit 0.005 ... Yes (0.006)
Checking if you have Authen::Passphrase 0.008 ... No
Checking if you have Minion::Backend::Redis 0.002 ... Yes (0.003)
Checking if you have URI 5.09 ... Yes (5.31)
Checking if you have App::Prove 1.7046 ... Yes (3.48)
Checking if you have Redis 1.995 ... Yes (2.000)
Checking if you have IO::Socket::SSL 2.067 ... Yes (2.089)
Checking if you have String::Similarity 1.04 ... Yes (1.04)
Checking if you have Cpanel::JSON::XS 4.06 ... Yes (4.39)
Checking if you have Net::DNS::Native 0.22 ... Yes (0.22)
==> Found dependencies: Authen::Passphrase
Searching Authen::Passphrase (0.008) on cpanmetadb ...
--> Working on Authen::Passphrase
Fetching https://www.cpan.org/authors/id/Z/ZE/ZEFRAM/Authen-Passphrase-0.008.tar.gz
-> OK
Unpacking Authen-Passphrase-0.008.tar.gz
Entering Authen-Passphrase-0.008
Checking configure dependencies from META.json
Checking if you have Module::Build 0.38 ... Yes (0.4234)
Checking if you have ExtUtils::Install 1.46 ... Yes (2.22)
Checking if you have warnings 0 ... Yes (1.70)
Checking if you have strict 0 ... Yes (1.13)
Configuring Authen-Passphrase-0.008
Running Build.PL
Checking prerequisites...
  requires:
    !  Crypt::DES is not installed

ERRORS/WARNINGS FOUND IN PREREQUISITES.  You may wish to install the versions
of the modules indicated above before proceeding with this installation

Created MYMETA.yml and MYMETA.json
Creating new 'Build' script for 'Authen-Passphrase' version '0.008'
-> OK
Checking dependencies from MYMETA.json ...
Checking if you have Digest 1.00 ... Yes (1.20)
Checking if you have Data::Entropy::Algorithms 0 ... Yes (0.007)
Checking if you have Params::Classify 0 ... Yes (0.015)
Checking if you have Crypt::Eksblowfish::Uklblowfish 0.008 ... Yes (0.009)
Checking if you have Module::Runtime 0.011 ... Yes (0.016)
Checking if you have Crypt::PasswdMD5 1.0 ... Yes (1.42)
Checking if you have Digest::SHA 6.02 ... Yes (6.04)
Checking if you have parent 0 ... Yes (0.241)
Checking if you have Digest::MD4 1.2 ... Yes (1.9)
Checking if you have Carp 0 ... Yes (1.54)
Checking if you have Crypt::DES 0 ... No
Checking if you have Crypt::Eksblowfish::Bcrypt 0.008 ... Yes (0.009)
Checking if you have warnings 0 ... Yes (1.70)
Checking if you have strict 0 ... Yes (1.13)
Checking if you have Authen::DecHpwd 2.003 ... Yes (2.007)
Checking if you have Crypt::UnixCrypt_XS 0.08 ... Yes (0.11)
Checking if you have Test::More 0 ... Yes (1.302199)
Checking if you have Crypt::MySQL 0.03 ... Yes (0.04)
Checking if you have MIME::Base64 2.21 ... Yes (3.16_01)
Checking if you have Digest::MD5 1.9953 ... Yes (2.58_01)
Checking if you have Module::Build 0 ... Yes (0.4234)

</details>

==> Found dependencies: Crypt::DES
Searching Crypt::DES (0) on cpanmetadb ...
--> Working on Crypt::DES
Fetching https://www.cpan.org/authors/id/D/DP/DPARIS/Crypt-DES-2.07.tar.gz
-> OK
Unpacking Crypt-DES-2.07.tar.gz
Entering Crypt-DES-2.07
Checking configure dependencies from META.yml
Checking if you have ExtUtils::MakeMaker 6.58 ... Yes (7.70)
Configuring Crypt-DES-2.07
Running Makefile.PL
Checking if your kit is complete...
Looks good
Generating a Unix-style Makefile
Writing Makefile for Crypt::DES
Writing MYMETA.yml and MYMETA.json
-> OK
Checking dependencies from MYMETA.json ...
Checking if you have ExtUtils::MakeMaker 0 ... Yes (7.70)
Building Crypt-DES-2.07
cp DES.pm blib/lib/Crypt/DES.pm
Running Mkbootstrap for DES ()
chmod 644 "DES.bs"
"/usr/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- DES.bs blib/arch/auto/Crypt/DES/DES.bs 644
"/usr/bin/perl" "/usr/share/perl5/core_perl/ExtUtils/xsubpp"  -typemap '/usr/share/perl5/core_perl/ExtUtils/typemap' -typemap '/home/[xxx]/.cpanm/work/1735662862.1758071/Crypt-DES-2.07/typemap'  DES.xs > DES.xsc
mv DES.xsc DES.c
cc -c   -D_REENTRANT -D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/include/db5.3 -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fexceptions -Wp,-D_FORTIFY_SOURCE=3 -Wformat -Werror=format-security -fstack-clash-protection -fcf-protection -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer -g -ffile-prefix-map=/build/perl/src=/usr/src/debug/perl -flto=auto   -DVERSION=\"2.07\" -DXS_VERSION=\"2.07\" -fPIC "-I/usr/lib/perl5/5.40/core_perl/CORE"   DES.c
DES.xs: In function ‘XS_Crypt__DES_expand_key’:
DES.xs:39:17: error: implicit declaration of function ‘perl_des_expand_key’; did you mean ‘_des_expand_key’? [-Wimplicit-function-declaration]
   39 |                 perl_des_expand_key((i8 *)key, ks);
      |                 ^~~~~~~~~~~~~~~~~~~
      |                 _des_expand_key
DES.xs: In function ‘XS_Crypt__DES_crypt’:
DES.xs:69:17: error: implicit declaration of function ‘perl_des_crypt’; did you mean ‘_des_crypt’? [-Wimplicit-function-declaration]
   69 |                 perl_des_crypt(input, SvGROW(output, output_len), (i32 *)ks, enc_flag);
      |                 ^~~~~~~~~~~~~~
      |                 _des_crypt
make: *** [Makefile:334: DES.o] Error 1
-> FAIL Installing Crypt::DES failed. See /home/[xxx]/.cpanm/work/1735662862.1758071/build.log for details. Retry with --force to force install it.
-> FAIL Installing the dependencies failed: Module 'Crypt::DES' is not installed
-> FAIL Bailing out the installation for Authen-Passphrase-0.008.
-> FAIL Installing the dependencies failed: Module 'Authen::Passphrase' is not installed
-> FAIL Bailing out the installation for ./tools/..

**Reply by Difegue** [reply]
I can see there was a similar bug raised in Gentoo - You might need to patch the Crypt::DES module before building it like they did.  https://bugs.gentoo.org/show_bug.cgi?id=870427

**Reply by UVJkiNTQ** [reply]
Yes, I have found the same bug report. It seems the case. I may try it. 
It is quite confused why upstream still keeps after the patch in gentoo in March 2023.

The bugzilla history shows:
dilfridge | 2024-10-25 20:48:26 UTC | Whiteboard |   | fixed in 2.70.0-r3

The build is using Crypt-DES-2.07 and cpan shows it is the latest version:
https://metacpan.org/pod/Crypt::DES

**Comment by UVJkiNTQ** [comment, marked as answer]
Marked as resolved since Archlinux repo provides the binary package for this perl module in [Extra] and this error therefore can be bypassed:

https://archlinux.org/packages/extra/x86_64/perl-crypt-des/

Others may add the module as dependancy in PKGBUILD to aviod this build error.

**Reply by UVJkiNTQ** [reply]
For others who found this thread:
The AUR package has been maintained.
